easybuild/framework/easyblock.py

```diff
@@ -1605,108 +1619,113 @@ def make_module_group_check(self):
-                        fullpath = os.path.join(self.installdir, pth)
-                        if os.path.isdir(fullpath) and dir_contains_files(fullpath, recursive=recursive):
-                            retained_paths.append(pth)
-                    if retained_paths != paths:
-                        self.log.info("Only retaining paths for %s that contain at least one file: %s -> %s",
-                                      key, paths, retained_paths)
-                        paths = retained_paths
-
-            if paths:
-                lines.append(self.module_generator.prepend_paths(key, paths))
+                mod_req_paths = []
+                top_level = getattr(self.module_load_environment, env_var).top_level_file
+                for path in search_paths:
+                    mod_req_paths.extend(self.expand_module_search_path(path, top_level, fake=fake))
+
+            if mod_req_paths:
+                mod_req_paths = nub(mod_req_paths)  # remove duplicates
+                mod_lines.append(self.module_generator.prepend_paths(env_var, mod_req_paths))
+
         if self.dry_run:
             self.dry_run_msg('')
 
-        if old_dir is not None:
-            change_dir(old_dir)
+        return "".join(mod_lines)
 
-        return ''.join(lines)
+    def expand_module_search_path(self, search_path, top_level, fake=False):
+        """
+        Expand given path glob and return list of suitable paths to be used as search paths:
+            - Paths are relative to installation prefix root
+            - Paths to files must exist and directories be non-empty
+            - Fake modules can set search paths to empty directories
+            - Search paths to a 'lib64' symlinked to 'lib' are discarded to avoid duplicates
+        """
+        # Expand globs but only if the string is non-empty
+        # empty string is a valid value here (i.e. to prepend the installation prefix root directory)
+        abs_glob = os.path.join(self.installdir, search_path)
+        exp_search_paths = [abs_glob] if search_path == "" else glob.glob(abs_glob)
+
+        # Explicitly check symlink state between lib dirs if it is still undefined (e.g. --module-only)
+        if self.install_lib_symlink == LibSymlink.NONE:
+            self.check_install_lib_symlink()
+
+        retained_search_paths = []
+        for abs_path in exp_search_paths:
+            # return relative paths
+            tentative_path = os.path.relpath(abs_path, start=self.installdir)
+            tentative_path = "" if tentative_path == "." else tentative_path  # use empty string instead of dot
+
+            # avoid duplicate entries between symlinked library dirs
+            tentative_sep = tentative_path + os.path.sep
+            if self.install_lib_symlink == LibSymlink.LIB64 and tentative_sep.startswith("lib64" + os.path.sep):
+                self.log.debug("Discarded search path to symlinked lib64 directory: %s", tentative_path)
+                break
```

Comment:
This should be `continue` rather than `break`, we should check all paths in `exp_search_paths` not jump out of the loop when we hit a symlink path?

Reply:
Indeed, using `break` here is wrong. It just happens to work because _luckily_ `glob` puts symlinks after folders in the expanded lists of paths.

Moreover, this method is currently untested. I am adding a new unit test for it to avoid such issues.